src/coreclr/jit/optimizer.cpp

```diff
@@ -4456,24 +4468,59 @@ void Compiler::optInvertWhileLoop(BasicBlock* block)
     // this is a conservative guess.
     if (auto copyFlags = bTest->bbFlags & (BBF_HAS_IDX_LEN | BBF_HAS_NULLCHECK | BBF_HAS_NEWOBJ | BBF_HAS_NEWARRAY))
     {
-        block->bbFlags |= copyFlags;
+        bNewCond->bbFlags |= copyFlags;
     }
 
-    /* Change the block to end with a conditional jump */
-
-    block->bbJumpKind = BBJ_COND;
-    block->bbJumpDest = bTest->bbNext;
+    bNewCond->bbJumpDest = bTest->bbNext;
+    bNewCond->inheritWeight(block);
 
-    /* Update bbRefs and bbPreds for 'block->bbNext' 'bTest' and 'bTest->bbNext' */
+    // Update bbRefs and bbPreds for 'bNewCond', 'bNewCond->bbNext' 'bTest' and 'bTest->bbNext'.
 
-    fgAddRefPred(block->bbNext, block);
+    fgAddRefPred(bNewCond, block);
+    fgAddRefPred(bNewCond->bbNext, bNewCond);
 
     fgRemoveRefPred(bTest, block);
-    fgAddRefPred(bTest->bbNext, block);
+    fgAddRefPred(bTest->bbNext, bNewCond);
+
+    // Move all predecessor edges that look like loop entry edges to point to the new cloned condition
+    // block, not the existing condition block. The idea is that if we only move `block` to point to
+    // `bNewCond`, but leave other `bTest` predecessors still pointing to `bTest`, when we eventually
+    // recognize loops, the loop will appear to have multiple entries, which will prevent optimization.
+    // We don't have loops yet, but blocks should be in increasing lexical numbered order, so use that
+    // as the proxy for predecessors that are "in" versus "out" of the potential loop. Note that correctness
+    // is maintained no matter which condition block we point to, but we'll lose optimization potential
+    // (and create spaghetti code) if we get it wrong.
+
+    BlockToBlockMap blockMap(getAllocator(CMK_LoopOpt));
+    bool            blockMapInitialized = false;
+
+    unsigned loopFirstNum  = bNewCond->bbNext->bbNum;
+    unsigned loopBottomNum = bTest->bbNum;
+    for (flowList* pred = bTest->bbPreds; pred != nullptr; pred = pred->flNext)
+    {
+        BasicBlock* predBlock = pred->getBlock();
+        unsigned    bNum      = predBlock->bbNum;
+        if ((loopFirstNum <= bNum) && (bNum <= loopBottomNum))
+        {
+            // Looks like the predecessor is from within the potential loop; skip it.
+            continue;
+        }
+
+        if (!blockMapInitialized)
+        {
+            blockMapInitialized = true;
+            blockMap.Set(bTest, bNewCond);
+        }
+
+        // Redirect the predecessor to the new block.
+        JITDUMP("Redirecting non-loop " FMT_BB " -> " FMT_BB " to " FMT_BB " -> " FMT_BB "\n", predBlock->bbNum,
+                bTest->bbNum, predBlock->bbNum, bNewCond->bbNum);
+        optRedirectBlock(predBlock, &blockMap, /*updatePreds*/ true);
```

Comment:
Since we expect `optRedirectBlock` to actually do something, I wonder if it should return a bool indicating that it made changes..?

Or, perhaps verify after all this redirecting that the block and the cloned block now have the expected numbers of preds?

Reply:
Hmmm... I think the only way it could fail to redirect is if the predecessor is BBJ_NONE, which  shouldn't occur. I suppose your suggestion is just to assert this. (Other types like BBJ_THROW, BBJ_RETURN, should never be predecessors.)

It turns out if it ends up not redirecting the pred, it's not fatal anyway, it's just sub-optimal.